File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/rewriting/PassThroughContentsRewriter.java
package com.codepoetics.raffia.streaming.rewriting;

import com.codepoetics.raffia.streaming.FilteringWriter;
import com.codepoetics.raffia.writers.BasketWriter;

import java.math.BigDecimal;

final class PassThroughContentsRewriter<T extends BasketWriter<T>> extends StreamingRewriter<T> {

  private int depth = 0;

  PassThroughContentsRewriter(T target, StreamingRewriter<T> parent) {
    super(target, parent, null);
  }

  @Override
  public FilteringWriter<T> end() {
    target = target.end();
    return depth-- == 0 ? parent.advance(target) : this;
  }

  @Override
  public FilteringWriter<T> beginObject() {
    target = target.beginObject();
    depth ++;
    return this;
  }

  @Override
  public FilteringWriter<T> beginArray() {
    target = target.beginArray();
    depth ++;
    return this;
  }

  @Override
  public FilteringWriter<T> key(String key) {
    return advance(target.key(key));
  }

  @Override
  public FilteringWriter<T> add(String value) {
    return advance(target.add(value));
  }

  @Override
  public FilteringWriter<T> add(BigDecimal value) {
    return advance(target.add(value));
  }

  @Override
  public FilteringWriter<T> add(boolean value) {
    return advance(target.add(value));
  }

  @Override
  public FilteringWriter<T> addNull() {
    return advance(target.addNull());
  }
}

File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/rewriting/PassThroughContentsRewriter.kt
package com.codepoetics.raffia.streaming.rewriting

import com.codepoetics.raffia.streaming.FilteringWriter
import com.codepoetics.raffia.writers.BasketWriter

import java.math.BigDecimal

internal class PassThroughContentsRewriter<T : BasketWriter<T>>(target: T, parent: StreamingRewriter<T>) : StreamingRewriter<T>(target, parent, null) {

    private var depth = 0

    override fun end(): FilteringWriter<T> {
        target = target.end()
        return if (depth-- == 0) parent!!.advance(target) else this
    }

    override fun beginObject(): FilteringWriter<T> {
        target = target.beginObject()
        depth++
        return this
    }

    override fun beginArray(): FilteringWriter<T> {
        target = target.beginArray()
        depth++
        return this
    }

    override fun key(key: String): FilteringWriter<T> {
        return advance(target.key(key))
    }

    override fun add(value: String): FilteringWriter<T> {
        return advance(target.add(value))
    }

    override fun add(value: BigDecimal): FilteringWriter<T> {
        return advance(target.add(value))
    }

    override fun add(value: Boolean): FilteringWriter<T> {
        return advance(target.add(value))
    }

    override fun addNull(): FilteringWriter<T> {
        return advance(target.addNull())
    }
}
